cmd/register/main.go

```diff
@@ -17,416 +17,171 @@ limitations under the License.
 
-	installDataMap, err := structToMap(config.Elemental.Install)
-	if err != nil {
-		return fmt.Errorf("failed to decode elemental-cli install data: %w", err)
+	// Use go-vfs afero compatibility layer (required by Viper)
+	afs := vfsafero.NewAferoFS(fs)
+	viper.SetFs(afs)
+	// Set final config path
+	log.Infof("Using base configuration file: %s", configPath)
+	viper.SetConfigFile(configPath)
+	// Merge config (considering bound flags)
+	if err := viper.MergeInConfig(); err != nil {
+		return fmt.Errorf("merging config: %w", err)
 	}
-
-	if err := elementalcli.Run(installDataMap); err != nil {
-		return fmt.Errorf("failed to install elemental: %w", err)
+	if err := viper.Unmarshal(&cfg); err != nil {
+		return fmt.Errorf("decoding configuration: %w", err)
 	}
-
 	return nil
 }
 
-func structToMap(str interface{}) (map[string]interface{}, error) {
-	var mapStruct map[string]interface{}
-
-	data, err := json.Marshal(str)
-	if err == nil {
-		if err := json.Unmarshal(data, &mapStruct); err == nil {
-			return mapStruct, nil
-		}
-	}
-
-	return nil, err
-}
-
-// isSystemInstalled checks if the host is currently installed
-// TODO: make the function dependent on tmp.Register returned data
-func isSystemInstalled() bool {
-	_, err := os.Stat(stateInstallFile)
-	return err == nil
-}
-
-func installRegistrationYAML(reg elementalv1.Registration) error {
-	registrationInBytes, err := yaml.Marshal(elementalv1.Config{
-		Elemental: elementalv1.Elemental{
-			Registration: reg,
-		},
-	})
-	if err != nil {
-		return err
-	}
-	f, err := os.Create(afterInstallHook)
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-
-	err = yaml.NewEncoder(f).Encode(schema.YipConfig{
-		Name: "Include registration config into installed system",
-		Stages: map[string][]schema.Stage{
-			"after-install": {
-				schema.Stage{
-					Directories: []schema.Directory{
-						{
-							Path:        filepath.Dir(registrationConf),
-							Permissions: 0700,
-						},
-					}, Files: []schema.File{
-						{
-							Path:        registrationConf,
-							Content:     string(registrationInBytes),
-							Permissions: 0600,
-						},
-					},
-				},
-			},
-		},
-	})
-
-	return err
-}
-
-func writeCloudInit(cloudConfig map[string]runtime.RawExtension) (string, error) {
-	f, err := os.CreateTemp(os.TempDir(), "*.yaml")
-	if err != nil {
-		return "", err
+func shouldSkipRegistration(stateHandler register.StateHandler, installer install.Installer) (bool, error) {
+	if !installer.IsSystemInstalled() {
+		return false, nil
 	}
-	defer f.Close()
-
-	bytes, err := util.MarshalCloudConfig(cloudConfig)
+	state, err := stateHandler.Load()
 	if err != nil {
-		return "", err
+		return false, fmt.Errorf("loading registration state")
 	}
-
-	log.Debugf("Decoded CloudConfig:\n%s\n", string(bytes))
-	_, err = f.Write(bytes)
-	return f.Name(), err
+	return !state.HasLastUpdateElapsed(registrationUpdateSuppressTimer), nil
```

Comment:
I just realized this only relies on the 24h delay time frame. Shouldn't we consider reboots as a valid case to register regardless of the timing of the last register? as context could have changed on a reboot.

Reply:
Opened a new issue: https://github.com/rancher/elemental-operator/issues/481